File: WordBreak.py
#All TC passed on leetcode


class Solution:
    def wordBreak(self, s: str, wordDict: List[str]) -> bool:

        #Here for every char position of s we check if its possible to split and create the string. If yes then it is stored as True in dp array.
         #Time complexity - O(n^3) - where n is length of s. n^2 for i and j iteration i.e. for every i we traverse j from 0 to i. Another O(n) is required for creating substrings in every iteration.
        #Space complexity - O(n)
        wordSet = set()
        for w in wordDict:
            wordSet.add(w)
        
        dp = [False]*(len(s)+1)
        dp[0] = True

        for i in range(len(s)+1):
            for j in range(i):
                if dp[j] and s[j:i] in wordSet:
                    dp[i] = True
                    break
        return dp[-1]


        

#----------------------------------------OR-----------------------------------------------
        #Time complexity - O(n^n) here for every letter we check if remanining letters are present one by one i.e for word=leetcode we check if 'l' is there then we check if 'e'->'ee'->'eet'->'eetc'->'eetco'->'eetcode' is present
        #Space complexity O(1)
        wordSet = set()
        for w in wordDict:
            wordSet.add(w)
        
        def helper(word, wordSet):
            if len(word)==0:
                return True
            
            for i in range(len(word)):
                sub = word[0:i+1]
                if sub in wordSet and helper(word[i+1:], wordSet):
                    return True
            return False

        return helper(s, wordSet)

        

File: src/WordBreak.java
// Time Complexity:  O(n^3)
// Space Complexity: O(n)

class Solution {

    Set<String> set;
    List<Integer> indexes;

    public boolean wordBreak(String s, List<String> wordDict) {
        
        set = new HashSet<>(wordDict);
        indexes = new ArrayList<>();
        indexes.add(0);

        for(int j=1; j<=s.length(); j++) {
            for(int k=0; k<indexes.size(); k++) {
                int i = indexes.get(k);                        // get all indexes from list where we found substring
                if(set.contains(s.substring(i, j))) {
                    indexes.add(j);                            // store all indexes where substring found in set
                    break;
                }
            }
        }
        
        int lastIndex = indexes.get(indexes.size()-1);

        return lastIndex == s.length();

    }
}



// // ******************** Another Method ********************

// // Time Complexity:  O(n^3)
// // Space Complexity: O(n)

// class Solution {

//     Set<String> set;
//     boolean[] dp;

//     public boolean wordBreak(String s, List<String> wordDict) {
        
//         set = new HashSet<>(wordDict);
//         dp = new boolean[s.length()+1];
//         dp[0] = true;

//         for(int j=1; j<dp.length; j++) {
//             for(int i=0; i<j; i++) {
//                 if(dp[i] && set.contains(s.substring(i, j))) {
//                     dp[j] = true;
//                     break;
//                 }
//             }
//         }

//         return dp[dp.length-1];

//     }
// }
